tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,519 @@ keywords:
+                int mid = i + size - 1;
+                int end = Math.min(i + 2 * size - 1, nums.length - 1);
+                merge2(nums, i, mid, end);
+            }
+        }
+    }
+    private void merge2(int[] nums, int l, int mid, int r) {
+        int[] tmp = new int[r - l + 1];
+        int i = l, j = mid + 1, k = 0;
+        while (i <= mid || j <= r) {
+            if (i > mid || j <= r && nums[i] > nums[j]) {
+                tmp[k++] = nums[j++];
+            } else {
+                tmp[k++] = nums[i++];
+            }
+        }
+        System.arraycopy(tmp, 0, nums, l, r - l + 1);
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time_Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
+
+**Space_Complexity** :  $O(n)$ 
+
+**Stablitiy**       :  Stable
+
+**Is In Place**      :  Not-In-Place
+
+**When to use**     :  
+1. When we don't have random access (linked list) , like as we have an array.
+2. When array is not to large
+
+  
+Merging of n elements takes n time and since each time the array is cut into half it takes   $\log_{2} n$   time to reach the top so **total time complexity is**    $O(n\log_{2} n)$  .
+
+
+![image](https://cdn.kastatic.org/ka-perseus-images/5fcbebf66560d8fc490de2a0d8a0e5b1d65c5c54.png)
+
+
+
+### **Mergesort is also useful for sorting linked lists**
+
+## [**Leetcode Problem 148_Sort List**](https://leetcode.com/problems/sort-list/)
+
+**Problem Statement** 
+
+`Given the head of a linked list, return the list after sorting it in ascending order.`
+
+**Test Cases**
+ 
+ *Test Case 1*
+
+```
+Input: head = [4,2,1,3]
+Output: [1,2,3,4]
+```
+![image](https://assets.leetcode.com/uploads/2020/09/14/sort_list_1.jpg)
+ 
+*Test Case 2*
+
+```
+Input: head = [-1,5,3,4,0]
+Output: [-1,0,3,4,5]
+```
+
+Here we can follow both top down or bottom up merge sort . I have already discussed the algorithms in detailed way for array .
+
+### **Intuition for the problem**
+
+```
+3->8->12->null
+1->10->23->null
+
+After merging we get :
+
+1->3->8->10->12->23->null.
+
+Of course the two lists must are ordered respectively.So we can only start to merge two lists that only 
+have one element,then we get a ordered list that have two element,do this again (that is "recursion").
+
+    Bottom Up Approach 
+
+    -1-> 5->  3-> 4->   0-> null
+    \  /       \ /         /
+   -1->5->null 0->3->4->null 
+      \        /         
+-1-> 0->3-> 4-> 5->null   (sorted) 
+
+    Top Down Approach
+
+    -1-> 5->  3-> 4-> 0-> null
+         /        \ 
+-1-> 5->null      3-> 4->0->null
+    /  \          /      \  
+-1->null 5->null 3->null  4->0->null
+     \   /        \        /     \
+    -1-> 5->null  3->null 4->null 0->null
+      \             \        \   /
+                 3->null    0-> 4->null
+        \              \      /  
+                    0->3->4->null
+          \          /
+        -1->0->3->4->5->null    (sorted)
+
+```
+
+
+**Leetcode Implementation of Linked List in cpp**
```

Comment:
What is the purpose of this code block ?

Reply:
I thought it would describe the implementation of LL shortly 

this block of code is given by leetcode
to understand the structure of LL @vigneshshiv

Reply:
@Sreetama2001 Add it as part of the code above or below, instead of separate block

Reply:
ohk